File: backend/app/schemas/game.py
"""创建游戏 / GameView 的 schema 与字段白名单（API_CONTRACT §6 §7 §8 §9 §10）。"""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

from ..core.errors import BadRequest

#: GameView **顶层字段白名单**（多一个都不允许，见 e2e_api.py 的 GAMEVIEW_KEYS）
GAMEVIEW_KEYS: tuple[str, ...] = (
    "game_id",
    "status",
    "revision",
    "viewer_player_id",
    "phase",
    "current_player",
    "decision_player",
    "observation",
    "public",
    "legal_actions",
    "events",
    "winner",
)

#: `public` 白名单（API_CONTRACT §7）
PUBLIC_KEYS: tuple[str, ...] = (
    "round",
    "deck_count",
    "discard_count",
    "players",
)

#: `public.players[]`（PlayerPublicView §8）白名单
PLAYER_PUBLIC_KEYS: tuple[str, ...] = (
    "player_id",
    "name",
    "alive",
    "hand_count",
    "is_current",
    "is_decision_player",
    "agent",
)

#: `observation` 白名单（API_CONTRACT §9 / INTERFACES §1.6）
OBSERVATION_KEYS: tuple[str, ...] = (
    "hand",
    "known_top",
    "actions_used",
    "max_actions_per_turn",
    "private_context",
)

#: 单条 LegalAction（API_CONTRACT §10）白名单
LEGAL_ACTION_KEYS: tuple[str, ...] = (
    "id",
    "type",
    "label",
    "enabled",
    "card_instance_id",
    "params",
)


class AgentRequest(BaseModel):
    """`POST /games` 里单个座位的 agent 配置。

    允许的字段（其余忽略，前端可能多传展示用字段）::

        {"type": "ismcts", "simulations": 500, "exploration": 1.4, "max_depth": 250}
        {"type": "mccfr", "model": "models/x.pkl"}  /  {"type": "mccfr", "path": "..."}
        {"type": "rule"} / {"type": "random"}
    """

    model_config = ConfigDict(extra="ignore")

    type: str = Field(default="rule", description="random | rule | ismcts | mccfr")
    simulations: Optional[int] = Field(default=None, ge=1)
    exploration: Optional[float] = Field(default=None, gt=0)
    max_depth: Optional[int] = Field(default=None, ge=1)
    model: Optional[str] = None
    path: Optional[str] = None

    def to_spec(self) -> dict:
        """转成 `agent_factory.build_agent` 认识的 dict（丢掉 None）。"""
        data = self.model_dump(exclude_none=True)
        return data


class CreateGameRequest(BaseModel):
    """`POST /games` 请求体（API_CONTRACT §6）。"""

    model_config = ConfigDict(extra="ignore")

    players: int = Field(default=3, ge=2, le=6)
    human_player: Optional[int] = Field(default=0, ge=0, le=5)
    agents: list[Optional[AgentRequest]] = Field(default_factory=list)
    seed: Optional[int] = None


def validate_create_request(req: CreateGameRequest) -> dict:
    """校验 §6 的三条规则，返回规范化后的 dict。

    规则::

        players == agents.length
        human_player 对应位置必须为 null
        其他座位必须指定 agent

    违反 → 400（参数错误）；schema 层面的错误由 FastAPI 自动给出 422。
    """
    players = int(req.players)
    agents = list(req.agents)

    if len(agents) != players:
        raise BadRequest(
            f"players({players}) 必须等于 agents 长度({len(agents)})",
            details={"players": players, "agents": len(agents)},
        )

    human = req.human_player
    if human is not None and not (0 <= int(human) < players):
        raise BadRequest(
            f"human_player({human}) 越界（0..{players - 1}）",
            details={"human_player": human, "players": players},
        )

    specs: list[Optional[dict]] = []
    for seat, entry in enumerate(agents):
        if entry is None:
            if human is None or seat != int(human):
                raise BadRequest(
                    f"座位 {seat} 缺少 agent 配置（只有 human_player 座位可以为 null）",
                    details={"seat": seat, "human_player": human},
                )
            specs.append(None)
            continue
        if human is not None and seat == int(human):
            raise BadRequest(
                f"座位 {seat} 是 human_player，agents[{seat}] 必须为 null",
                details={"seat": seat, "human_player": human},
            )
        specs.append(entry.to_spec())

    return {
        "players": players,
        "human_player": None if human is None else int(human),
        "agent_specs": specs,
        "seed": req.seed,
    }


__all__ = [
    "GAMEVIEW_KEYS",
    "PUBLIC_KEYS",
    "PLAYER_PUBLIC_KEYS",
    "OBSERVATION_KEYS",
    "LEGAL_ACTION_KEYS",
    "AgentRequest",
    "CreateGameRequest",
    "validate_create_request",
]

File: frontend/src/types/game.ts
/**
 * 对局视图与动作类型 —— 逐字段对齐 API_CONTRACT.md §3/§6/§7/§8/§9/§10/§11/§12。
 * 后端是唯一规则权威：前端只渲染 view.legal_actions 并回传 action.id + payload。
 */
import type {
  AgentSpec,
  CardInstance,
  CardSpec,
  KnownTopCard,
  PrivateContext,
} from './card';
import type { GameEvent } from './event';

export type Phase = 'ACTION' | 'COUNTER' | 'REORDER' | 'REINSERT' | 'ENDED';

export type GameStatus = 'playing' | 'ended' | 'aborted' | string;

/** legal_actions[].type
 *  `ESCAPE` 是**规则改动新增**的反应动作（INTERFACES A13）：新规则下遁术只能在反制窗口打出，
 *  与 `COUNTER` / `PASS_COUNTER` 并列出现在同一窗口（旧规则里遁术是行动阶段的 `PLAY_CARD`）。 */
export type ActionType =
  | 'END_ACTION'
  | 'PLAY_CARD'
  | 'PLAY_CARD_TARGET'
  | 'COUNTER'
  | 'PASS_COUNTER'
  | 'ESCAPE'
  | 'REORDER_TOP'
  | 'REINSERT_TRIBULATION';

export type ReinsertRegion = 'TOP' | 'NEAR_TOP' | 'MIDDLE' | 'BOTTOM';

export const REINSERT_REGIONS: readonly ReinsertRegion[] = [
  'TOP',
  'NEAR_TOP',
  'MIDDLE',
  'BOTTOM',
] as const;

export interface EnumParam<T = string | number> {
  type: 'enum';
  options: T[];
}

export interface TokenOrderParam {
  type: 'token_order';
  /** 可选：后端若显式给出 token 列表，前端据此渲染 */
  options?: string[];
}

export interface ActionParams {
  target_player?: EnumParam<number>;
  region?: EnumParam<ReinsertRegion>;
  order?: TokenOrderParam;
  [key: string]: unknown;
}

/** API_CONTRACT §10：整个 API 最重要的数据 */
export interface LegalAction {
  id: string;
  type: ActionType;
  label: string;
  enabled?: boolean;
  card_instance_id?: string | null;
  params?: ActionParams | null;
}

/** POST /games/{id}/actions 的 payload：按动作类型给不同键 */
export interface ActionPayload {
  target_player?: number;
  order?: string[];
  region?: ReinsertRegion;
  [key: string]: unknown;
}

export interface SubmitActionRequest {
  revision: number;
  action_id: string;
  payload?: ActionPayload;
}

export interface Observation {
  hand: CardInstance[];
  known_top: KnownTopCard[];
  actions_used: number;
  max_actions_per_turn: number;
  private_context: PrivateContext | null;
}

/** API_CONTRACT §8：严禁包含真实手牌 / 私有观星结果 */
export interface PlayerPublicView {
  player_id: number;
  name: string;
  alive: boolean;
  hand_count: number;
  is_current: boolean;
  is_decision_player: boolean;
  agent?: AgentSpec | null;
  avatar?: string | null;
}

export interface PublicState {
  round: number;
  deck_count: number;
  discard_count: number;
  players: PlayerPublicView[];
  /** 契约未固定，后端可给：弃牌堆最上一张 / 当前回合数 */
  last_discard?: string | null;
  turn_no?: number;
}

/** API_CONTRACT §7：创建 / 读取 / 执行动作统一返回 GameView */
export interface GameView {
  game_id: string;
  status: GameStatus;
  revision: number;
  viewer_player_id: number;
  phase: Phase;
  current_player: number;
  decision_player: number;
  observation: Observation;
  public: PublicState;
  legal_actions: LegalAction[];
  events: GameEvent[];
  /** 以下为后端可选补充字段 */
  winner?: number | null;
  forced_stop?: boolean;
  created_at?: string;
  error?: string | null;
}

/** POST /games */
export interface CreateGameRequest {
  players: number;
  human_player: number;
  agents: (AgentSpec | null)[];
  seed?: number | null;
}

export interface CreateGameResponse {
  game_id: string;
  player_id: number;
  state: GameView;
}

export interface HealthResponse {
  status: string;
  version: string;
}

/** GET /cards（API_CONTRACT §18） */
export interface CardsResponse {
  cards: CardSpec[];
}

/** DELETE /games/{id} */
export interface DeleteGameResponse {
  ok: boolean;
}

/** GET /health 之外的通用错误体（API_CONTRACT §3） */
export interface ApiErrorBody {
  error: {
    code: string;
    message: string;
    details?: Record<string, unknown> | null;
  };
}
